daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ClosedSubstitution.scala

```diff
@@ -0,0 +1,43 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf
+package engine
+package preprocessing
+
+import data.Ref
+
+// To handle a closed substitution in a lazy way while traversing a term top down.
+private[preprocessing] class ClosedSubstitution[X] private (
+    mapping: Map[Ref.Name, (ClosedSubstitution[X], X)],
+) {
+
+  def get(name: Ref.Name): Option[(ClosedSubstitution[X], X)] =
+    mapping.get(name)
+
+  def introVar(name: Ref.Name, x: X) =
+    new ClosedSubstitution(mapping.updated(name, this -> x))
+
+  // requirement: names.size == xs.size
+  def introVars(names: Iterable[Ref.Name], xs: Iterable[X]): ClosedSubstitution[X] = {
```

Comment:
Are we passing the two separately rather than the zipped version purely for performance reasons? The zipped version seems a bit nicer otherwise.

Reply:
I changed to the zipped version.